pkg/hints/hints.go

```diff
@@ -0,0 +1,155 @@
+		largeTableThreshold:  largeTableThreshold,
+	}
+
+	engineInstance := reflect.ValueOf(he)
+	engineType := reflect.TypeOf(he)
+	for i := 0; i < engineType.NumMethod(); i++ {
+		method := engineType.Method(i)
+		boundMethod := engineInstance.Method(i)
+		if processor, ok := boundMethod.Interface().(func(dba.DatabaseMigrationSchemaHint) ([]dba.ManagedDatabaseError, error)); ok {
+			he.operationToProcessor[strcase.ToLowerCamel(method.Name)] = processorFunc(processor)
+		}
+	}
+
+	return he
+}
+
+type processorFunc func(dba.DatabaseMigrationSchemaHint) ([]dba.ManagedDatabaseError, error)
+
+type hintsEngine struct {
+	handle              dbadmin.DbAdmin
+	largeTableThreshold uint64
+
+	operationToProcessor map[string]processorFunc
+}
+
+func (he *hintsEngine) ProcessHints(hints []dba.DatabaseMigrationSchemaHint) ([]dba.ManagedDatabaseError, error) {
+	var migrationErrors []dba.ManagedDatabaseError
+	for _, hint := range hints {
+		processor, ok := he.operationToProcessor[hint.Operation]
+		if !ok {
+			unknownOp := newMigrationErrorF("Unknown hint operation: %s", hint.Operation)
+			migrationErrors = append(migrationErrors, unknownOp)
+		} else {
+			processorErrors, err := processor(hint)
+			if err != nil {
+				return migrationErrors, err
+			}
+			migrationErrors = append(migrationErrors, processorErrors...)
+		}
+	}
+
+	return migrationErrors, nil
+}
+
+func (he *hintsEngine) AddColumn(hint dba.DatabaseMigrationSchemaHint) (migrationErrors []dba.ManagedDatabaseError, err error) {
+	var estimates map[dbadmin.TableName]uint64
+
+	tableName := dbadmin.TableName(hint.TableName)
+	estimates, err = he.handle.GetTableSizeEstimates([]dbadmin.TableName{tableName})
+	if err != nil {
+		return
+	}
+
+	tableSizeEstimate := estimates[tableName]
+	if tableSizeEstimate > he.largeTableThreshold {
+		migrationErrors = append(migrationErrors, newMigrationErrorF(addColumnTableTooLarge, tableName))
+	}
+
+	for _, column := range hint.Columns {
+		if column.NotNullable && !column.HasServerDefault {
+			migrationErrors = append(migrationErrors, newMigrationErrorF(addNotNullableColumnNoDefault, column.Name, tableName))
+		}
+	}
+
+	return
+}
+
+func (he *hintsEngine) CreateIndex(hint dba.DatabaseMigrationSchemaHint) (migrationErrors []dba.ManagedDatabaseError, err error) {
+	tableName := dbadmin.TableName(hint.TableName)
+
+	switch hint.IndexType {
+	case "unique":
+		columnNames := make([]string, len(hint.Columns))
+		for _, columnReference := range hint.Columns {
+			columnNames = append(columnNames, columnReference.Name)
+		}
+
+		var willFail bool
+		willFail, err = he.handle.ConstraintWillFail(tableName, dbadmin.UniqueConstraint, columnNames...)
+		if err != nil {
+			return
+		}
+		if willFail {
+			migrationErrors = append(migrationErrors, newMigrationErrorF(addUniqueIndexWillFail, tableName, columnNames))
+		}
+	case "index":
+		return migrationErrors, fmt.Errorf("Not implemented yet for index type: %s", hint.IndexType)
```

Comment:
Maybe add TODOs? My editor marks them as warnings, so it's a little valuable to me, idk how you feel about them.

Reply:
Actually I can't ignore these, because if anybody specifies such a hint it will pause the migration. I can either say they're all OK, or I can implement the logic.

Reply:
I didn't mean to suggest removing the return statement, rather just adding an TODO comment above with some context.

Reply:
I'm saying I need to remove the return statements and either replace them with noop/no-error, or actually implement the logic now. Otherwise any hint that says it's creating a normal index is going to pause the migration when the hints engine throws an error.

Reply:
Added some logic here.